merkledag/utils/utils.go

```diff
@@ -0,0 +1,113 @@
+package dagutils
+
+import (
+	"errors"
+	"time"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+func AddLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname string, childk key.Key) (*dag.Node, error) {
+	if childname == "" {
+		return nil, errors.New("cannot create link with no name!")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+	childnd, err := ds.Get(ctx, childk)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure no link with that name already exists
+	_ = root.RemoveNodeLink(childname) // ignore error, only option is ErrNotFound
+
+	err = root.AddNodeLinkClean(childname, childnd)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = ds.Add(root)
+	if err != nil {
+		return nil, err
+	}
+	return root, nil
+}
+
+func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key, create bool) (*dag.Node, error) {
```

Comment:
> paths separated by a given delimiter are a unixfs thing

no, this is part of the dag. path traversal is why ipfs has human readable string names in the links.

> @jbenet is this something you want in this PR? sure it would be nice, but its a bit much for a PR about adding an extra flag to object patch.

yes because this is what patch is, this data structure is what the patch command is doing. it will make every other patch command easier to reason about and call. it will make things like the write-and-update-merkle-root part of ipns fs easier to liberate and use with just ipfs objects (not ipns).